commit 7d15f156a60e2e32f25ab3a7a9abc0fbbfe011d6 (refs/changes/73/2873/3)
Author: Jan Wyszynski <jan.wyszynski@joyent.com>
Date:   2017-10-26T18:32:50+00:00 (1 year, 11 months ago)
    
    MANTA-3475 muskie should report 503 errors when moray hits its max queue length

diff --git a/lib/errors.js b/lib/errors.js
index 8e1e22d..485d1a9 100644
--- a/lib/errors.js
+++ b/lib/errors.js
@@ -574,11 +574,12 @@ function ResourceNotFoundError(p) {
 util.inherits(ResourceNotFoundError, MuskieError);
 
 
-function ServiceUnavailableError(req) {
+function ServiceUnavailableError(req, cause) {
     MuskieError.call(this, {
         restCode: 'ServiceUnavailable',
         statusCode: 503,
-        message: 'manta is unable to serve this request'
+        message: 'manta is unable to serve this request',
+        cause: cause
     });
 }
 util.inherits(ServiceUnavailableError, MuskieError);
diff --git a/lib/server.js b/lib/server.js
index b04eb4b..6ccf254 100644
--- a/lib/server.js
+++ b/lib/server.js
@@ -12,6 +12,7 @@ var crypto = require('crypto');
 var fs = require('fs');
 var path = require('path');
 var url = require('url');
+var verror = require('verror');
 
 var assert = require('assert-plus');
 var bunyan = require('bunyan');
@@ -47,6 +48,10 @@ var JOBS_LIVE_RE = /(state|status|name)=/i;
 // Always force JSON
 function formatJSON(req, res, body) {
     if (body instanceof Error) {
+        var info = verror.info(body);
+        if (info.morayMaxQueueDepthReached) {
+            body = new ServiceUnavailableError(req, body);
+        }
         body = translateError(body, req);
         res.statusCode = body.statusCode || 500;
         if (res.statusCode >= 500)
@@ -185,26 +190,33 @@ function createServer(options, clearProxy) {
 
     server.use(function ensureDependencies(req, res, next) {
         var ok = true;
+        var errors = [];
 
         if (!options.picker()) {
+            errors.push(new Error('picker unavailable'));
             req.log.error('picker unavailable');
             ok = false;
         } else if (!options.moray()) {
+            errors.push(new Error('index moray unavailable'));
             req.log.error('index moray unavailable');
             ok = false;
         } else if (!options.mahi()) {
+            errors.push(new Error('mahi unavailable'));
             req.log.error('mahi unavailable');
             ok = false;
         } else if (!options.marlin()) {
+            errors.push(new Error('marlin available'));
             req.log.error('marlin unavailable');
             ok = !req.isMarlinRequest();
         } else if (!options.medusa()) {
+            errors.push(new Error('medusa unavailable'));
             req.log.error('medusa unavailable');
             ok = !req.isMedusaRequest();
         }
 
         if (!ok) {
-            next(new ServiceUnavailableError());
+            next(new ServiceUnavailableError(req,
+                        new verror.MultiError(errors)));
         } else {
             next();
         }
diff --git a/test/errors.test.js b/test/errors.test.js
new file mode 100644
index 0000000..76c7fbd
--- /dev/null
+++ b/test/errors.test.js
@@ -0,0 +1,34 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright (c) 2017, Joyent, Inc.
+ */
+
+var _helper = __dirname + '/helper.js';
+if (require.cache[_helper])
+    delete require.cache[_helper];
+var helper = require(_helper);
+
+var util = require('util');
+var verror = require('verror');
+
+var errors = require('../lib/errors');
+
+var test = helper.test;
+
+test('ServiceUnavailableError', function (t) {
+    var cause = new Error('cause');
+    var err = new errors.ServiceUnavailableError(null, cause);
+
+    t.ok(err instanceof errors.MuskieError, 'error is a MuskieError');
+    t.equal(err.restCode, 'ServiceUnavailable');
+    t.equal(err.statusCode, 503);
+    t.equal(err.message, 'manta is unable to serve this request');
+    t.deepEqual(err.cause(), cause);
+
+    t.end();
+});
